src/main/kotlin/io/emeraldpay/dshackle/upstream/UpstreamRpcMethodsDetector.kt

```diff
@@ -0,0 +1,55 @@
+package io.emeraldpay.dshackle.upstream
+
+import io.emeraldpay.dshackle.config.UpstreamsConfig
+import io.emeraldpay.dshackle.upstream.rpcclient.CallParams
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+import reactor.core.publisher.Mono
+
+typealias UpstreamRpcMethodsDetectorBuilder = (Upstream, UpstreamsConfig.Upstream<*>?) -> UpstreamRpcMethodsDetector?
+
+abstract class UpstreamRpcMethodsDetector(
+    private val upstream: Upstream,
+    private val config: UpstreamsConfig.Upstream<*>? = null,
+) {
+    protected val log: Logger = LoggerFactory.getLogger(this::class.java)
+
+    open fun detectRpcMethods(): Mono<Map<String, Boolean>> = detectByMagicMethod().switchIfEmpty(detectByMethod())
+
+    protected fun detectByMethod(): Mono<Map<String, Boolean>> =
+        Mono.zip(
+            rpcMethods().map {
+                Mono
+                    .just(it)
+                    .flatMap { (method, param) ->
+                        upstream
+                            .getIngressReader()
+                            .read(ChainRequest(method, param))
+                            .flatMap(ChainResponse::requireResult)
+                            .map { method to true }
+                            .onErrorResume { err ->
+                                val notAvailableError =
+                                    listOf(
+                                        "method ([A-Za-z0-9_]+) does not exist/is not available",
```

Comment:
Let's initialize the list of regexps once. We don't need to do such operation each time we encounter an error